### PR TITLE
Ensure unique constraint names in MLflow tracking database

### LINKED ISSUE #1229: Bug: Check Constrain Name Duplicate Name ( breaks MS SQL )
Hi,

we have started using ML Flow 0.9.1 with MS SQL service and it looks like SCHEMA generate by mlflow has duplicate names for check constrains in `runs` and `experiments` tables.

Error log :

```
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42S01', "[42S01] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]There is already an object named 'lifecycle_stage' in the database. (2714) (SQLExecDirectW)")
[SQL: 
CREATE TABLE runs (
 run_uuid VARCHAR(32) NOT NULL, 
 name VARCHAR(250) NULL, 
 source_type VARCHAR(20) NULL, 
 source_name VARCHAR(500) NULL, 
 entry_point_name VARCHAR(50) NULL, 
 user_id VARCHAR(256) NULL, 
 status VARCHAR(20) NULL, 
 start_time BIGINT NULL, 
 end_time BIGINT NULL, 
 source_version VARCHAR(50) NULL, 
 lifecycle_stage VARCHAR(20) NULL, 
 artifact_uri VARCHAR(200) NULL, 
 experiment_id INTEGER NULL, 
 CONSTRAINT run_pk PRIMARY KEY (run_uuid), 
 CONSTRAINT source_type CHECK (source_type IN ('NOTEBOOK', 'JOB', 'LOCAL', 'UNKNOWN', 'PROJECT')), 
 CONSTRAINT status CHECK (status IN ('SCHEDULED', 'FAILED', 'FINISHED', 'RUNNING')), 
 CONSTRAINT lifecycle_stage CHECK (lifecycle_stage IN ('active', 'deleted')), 
 FOREIGN KEY(experiment_id) REFERENCES experiments (experiment_id)
)
]
```

MS SQL doesn't support constrains name duplicates across the SCHEMA/DB.

it can be mitigated by renaming constrains in lines 61 and 139 to include a table name prefix.

https://github.com/mlflow/mlflow/blob/fcc87c4965e2141fff996826b5e2f9780a996059/mlflow/store/dbmodels/initial_models.py#L61

https://github.com/mlflow/mlflow/blob/fcc87c4965e2141fff996826b5e2f9780a996059/mlflow/store/dbmodels/initial_models.py#L139

### PR DESCRIPTION
**Pull request by smurching**
## What changes are proposed in this pull request?
Fixes https://github.com/mlflow/mlflow/issues/1229 

The MLflow tracking database currently has multiple constraints with the same name. Several DBMSes (e.g. MySQL >= 8.0, MS SQL server) require unique constraint names, and so running an MLflow server with SQL-backed storage fails during table creation against these DBMSes. This PR updates the MLflow tracking database schema to contain unique constraint names, adding a database migration to accomplish this for existing users of SQL-backed storage.
 
## How is this patch tested?
Manually ran the MySQL tests against a MySQL 8.0 database (via `MYSQL_TEST_USERNAME=<my-username> MYSQL_TEST_PASSWORD=<my-password> pytest tests/store/test_sqlalchemy_store.py::TestSqlAlchemyStoreMysqlDb --release`) and verified that this now passes but failed prior to this PR. Did not otherwise add new tests - see below for context: 

After this PR & the MLflow 1.0 release, users can have a valid MLflow database by either:
* Creating their DB before 1.0, upgrading to MLflow 1.0, and running `mlflow db upgrade`, which will rename constraints to eliminate duplicate names and run the metric step migration from #1155. The upgrade logic for pre-1.0 users will be deprecated and removed in releases after 1.0, so pre-1.0 users must upgrade specifically to MLflow 1.0 and run `mlflow db upgrade` before they can use later versions (e.g. MLflow 1.1).
* Creating their DB after 1.0, in which case we initialize it via the schema in [initial_models.py](https://github.com/smurching/mlflow/blob/db4eba97b65c4255ef347812510306493f613088/mlflow/store/dbmodels/initial_valid_models.py) and run the metric step migration from #1155

Existing tests already verify case a) by running SQLAlchemyStore tests against a migrated database and case b) by running SQLAlchemyStore tests against a database created from scratch. Moreover, we also have tests verifying that case a) and case b) generate DBs with the same schema, so the (identical) behavior of both cases should already be guaranteed by tests.
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Fixes an issue where the MLflow tracking database had multiple constraints with the same name, thereby enabling running an MLflow tracking server against a DBMS that requires unique constraint names (e.g. MySQL >= 8.0).
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [x] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/sqlalchemy_store.py [hunk 1]
```diff
@@ -95,22 +94,30 @@ def __init__(self, db_uri, default_artifact_root):
...
-                              "and desired schema, exiting. Diff:\n %s",
-                              pprint.pformat(diff, indent=2, width=20))
+            current_rev = mc.get_current_revision()
+            if current_rev != head_revision:
                 raise MlflowException(
-                    "Detected out-of-date database schema. Take a backup of your database, then "
-                    "run 'mlflow db upgrade %s' to migrate your database to the latest schema. "
-                    "NOTE: schema migration may result in database downtime "
-                    "- please consult your database's documentation for more detail." % engine.url)
+                    "Detected out-of-date database schema (found version %s, but expected %s)."
```
Comment (dbczumar) [inline mlflow/store/sqlalchemy_store.py]: Nit: Missing space between the period and `Take` in the next sentence:

```
mlflow.exceptions.MlflowException: Detected out-of-date database schema (found version 451aebb31d03, but expected ff01da956556).Take a backup of your database, then run 'mlflow db upgrade sqlite:///mydb' to migrate your database to the latest schema. NOTE: schema migration may result in database downtime - please consult your database's documentation for more detail.
```

**Review by dbczumar — CHANGES_REQUESTED** [review]
@smurching Left a couple nits and small documentation comments. As we discussed offline, there appears to be a migration ordering issue that we need to address prior to merging.

File: mlflow/temporary_db_migrations_for_pre_1_users/alembic.ini [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# A generic, single database configuration.
```
Comment (smurching) [inline mlflow/temporary_db_migrations_for_pre_1_users/alembic.ini]: This whole directory is a copy of the existing `alembic` directory, except containing a different set of migrations (the migration for pre-1.0 users) under the `versions` subdirectory

File: mlflow/temporary_db_migrations_for_pre_1_users/versions/ff01da956556_ensure_unique_constraint_names.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
+"""ensure_unique_constraint_names
```
Comment (smurching) [inline mlflow/temporary_db_migrations_for_pre_1_users/versions/ff01da956556_ensure_unique_constraint_names.py]: This is the one-time migration for pre-1.0 users to rename constraints with duplicate names

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @smurching. I tested the upgrade flow from `v0.9.1` and ensured that adding subsequent migrations does not break the `mlflow db upgrade` or fresh DB initialization flow.

**Pull request merged**